tests/testthat/test-frgtracker.R

```diff
@@ -19,6 +16,86 @@ test_that("multiplication works", {
 
 #function5 start
 
+generate_fake_dataframe <- function(){
+  df <- data.frame (course_id  = c(rep("511", 4), rep("522", 4)),
+                    student_id =  c("tom", "tiff","mike","joel",
+                                    "tom","tiff","mike","joel"),
+                    grade = c(90, 70, 60, 50, 95, 91, 85, 80)
+  )
+  df
+}
+
+
+generate_fake_dataframe_output_511_course <- function(){
+  df <- data.frame (student_id = c("tom", "tiff","mike","joel"),
+                    grade = c(90, 70, 60, 50),
+                    rank = c(1, 2, 3, 4)
+  )
+  df
+}
+
+generate_fake_dataframe_output_522_course <- function(){
+  df <- data.frame (student_id = c("tom", "tiff","mike","joel"),
+                    grade = c(95, 91, 85, 80),
+                    rank = c(1, 2, 3, 4)
+  )
+  df
+}
+
+generate_fake_dataframe_output_all_course <- function(){
+  df <- data.frame (student_id = c("tom", "tiff","mike","joel"),
+                    grade = c(92.5, 80.5, 72.5, 65),
+                    rank = c(1, 2, 3, 4)
+  )
+  df
+}
+
+
+generate_fake_matrix <- function(){
+  matrix_ex <- matrix(data = c(c(rep("511",4), rep("522", 4)),
+                               c("tom", "tiff","mike","joel","tom",
+                                 "tiff","mike","joel"),
+                               c(84.66, 88.34, 87.66, 90.82, 95.52,
+                                 87.92, 88.92, 92.80)),
+                      nrow = 8,
+                      ncol = 3)
+}
+
+
+test_that("Incorrect input types should throw an error", {
+  temp_df = generate_fake_dataframe()
+
+  expect_error(rank_students(df = generate_fake_matrix()))
+  expect_error(rank_students(df = temp_df, courseid = 512))
+  expect_error(rank_students(df = temp_df, n = "3"))
+  expect_error(rank_students(df = temp_df, n = -3))
+  expect_error(rank_students(df = temp_df, n = 3.5))
+  expect_error(nrow(rank_students(df = temp_df,courseid="511", n = 5)))
+  expect_error(rank_students(df = temp_df, ascending = "TRUE"))
+})
+
+test_that("Dataframe should be equal", {
+  temp_df = generate_fake_dataframe()
+  expect_equal(rank_students(df = temp_df,courseid="511"),
+                   generate_fake_dataframe_output_511_course())
+  expect_equal(rank_students(df = temp_df,courseid="522"),
+               generate_fake_dataframe_output_522_course())
+  expect_equal(rank_students(df = temp_df,courseid="all"),
+               generate_fake_dataframe_output_all_course())
+})
+
+test_that("Grade should be between 0 and 100", {
+  temp_df = generate_fake_dataframe()
+  expect_false(any(temp_df$grade < 0))
+  expect_false(any(temp_df$grade > 100))
+})
+
+test_that("NAs should be dropped", {
+  temp_df = generate_fake_dataframe()
+
+  expect_false(is.null(rank_students(df = temp_df)))
+})
+
 #function5 end
```

Comment:
can you rename the function name here as well?

Reply:
done!